# Here, CAF and Macro subpopulations are analyzed through scRNA-seq and spatial transcriptomics.
### scRNA CAF
```
load("analysis/PDAC_TME_sc.Rdata")
table(PDAC_TME_sc$celltype)
Fibro <- subset(PDAC_TME_sc, celltype == "Fibroblast")
Fibro <- RunPCA(Fibro)
Fibro <- RunHarmony(Fibro,group.by.vars = "sample")
#PDAC_scRNA <- JackStraw(PDAC_scRNA, num.replicate = 10,reduction = "pca",assay = "RNA")
ElbowPlot(Fibro,reduction = "harmony")

Fibro <- RunUMAP(Fibro,reduction = "harmony",dims = 1:30)
Fibro <- FindNeighbors(Fibro, reduction = "harmony", dims = 1:30)
Fibro <- FindClusters(Fibro,
                      resolution =0.3,
                      method ='igraph',
                      verbose = T)# 6
#defined celltype subtype
table(Fibro$SCT_snn_res.0.3)
cluster<- c("Fibro_c1",
            "Fibro_c2",
            "Fibro_c3",
            "Fibro_c4",
            "Fibro_c5",
            "Fibro_c6"
            )
new.cluster.ids <- cluster
names(new.cluster.ids)<- levels(Fibro)
Fibro <- RenameIdents(Fibro, new.cluster.ids)
Fibro@meta.data$celltype_subtype<-Fibro@active.ident

table(Fibro@active.ident)
Fibro$site <- factor(Fibro$site,levels = c("Normal pancreas",'Adjacent',"Liver metastases","PDAC primary"))
#UMAP plot
#Fig 3A
p<-DimPlot(Fibro,group.by = "celltype_subtype",cols = col1[1:6],raster = T,split.by = "site")
pdf("/result/scRNA/TME/Figure5_Fibro_Macro/Figure_5_Fibro_cellcluster_UMAP_legend.pdf",width = 21,height = 6)
print(p)
dev.off()

#Dotplot
#Fig S3G
markers<-c("IL6",'PDGFRA','CXCL12','CFD','CXCL1',#iCAF
           "ACTA2",'MMP11','MYL9','HOPX','TPM1',#myCAF
           "HLA-B",'SAA3','CD74'#apCAF antigen presenting CAFs
           )
p <- DotPlot(Fibro,features = markers,group.by = "celltype_subtype")+scale_color_gradient2(low = 'white',mid = col4[50],high = col4[20])+
  theme_bw()+theme(panel.grid.major = element_blank() ,axis.text.x = element_text(size = 15,color = "black",angle = 45,hjust = 1,vjust = 1),
                   axis.text.y = element_text(size = 15,color = "black"),
                   axis.title = element_text(size = 15,color = 'black'),
                   legend.text = element_text(size = 12,color = "black"))+labs(x="",y="")
p
pdf("/result/scRNA/TME/Figure5_Fibro_Macro/Figure_5_Fibro_cellcluster_Marker.pdf",width = 6,height = 6)
print(p)
dev.off()


#percent
#Fig 3B
d6 <- dittoBarPlot(
  object =Fibro ,
  var = "celltype_subtype",color.panel = col1[1:6],x.reorder = c(3,1,2,4),
  #colors = c(3,5,6,4,1,2),
  group.by = "site")
d6<-d6+theme(axis.text.x = element_text(size = 15,color = "black",angle = 45,hjust = 1,vjust = 1),
             axis.text.y = element_text(size = 15,color = "black"),
             axis.title = element_text(size = 15,color = 'black'),
             legend.text = element_text(size = 12,color = "black"))

pdf("result/scRNA/TME/Figure5_Fibro_Macro/Figure_6_Fibro_percent_cellcluster.pdf",width = 4,height = 4)
print(d6)
dev.off()

marker_Fibro <- FindAllMarkers(Fibro,
                               assay = "SCT",
                               test.use = "wilcox",
                               logfc.threshold = 0.5,
                               #min.diff.pct = 0.3,
                               only.pos = T
                               )
write.table(marker_Fibro,"result/scRNA/TME/Figure5_Fibro_Macro/marker_Fibro.txt",sep = "\t",quote = F)

markers_cluster <- marker_Fibro
top10<-markers_cluster %>% group_by(cluster) %>% top_n(n=10,wt=avg_log2FC)
head(top10)

gmt <- matrix("",12,length(cluster))
for (i in 1:length(cluster)){
  gmt[1,] <- cluster
  #gmt[2,] <-""
  gmt[3:12,i]<-top10$gene[(10*(i-1)+1):(10*i)]
  
}
gmt <- as.data.frame(t(gmt))
write.table(gmt,"analysis/Bulk-analysis/celltype_subtype/Fibro_gmt.gmt",sep = "\t",row.names = F,col.names = F,quote = F)
#


top10<-top10[!duplicated(top10$gene),]
p<-DotPlot(object = Fibro,
           features = top10$gene,
           assay = "SCT")+coord_flip()+theme(axis.text.x=element_text(angle=45,size=8))
pdata<-p$data
top10_avgexp<-pdata[,c(1,3,4)]
top10_avgexp<-na.omit(top10_avgexp)

top10_avgexp<-dcast(top10_avgexp,features.plot~id,value.var = 'avg.exp')
rownames(top10_avgexp)<-top10_avgexp$features.plot
top10_avgexp<-top10_avgexp[,-1]
top10_avgexp<-top10_avgexp[rowSums(top10_avgexp)>0,]
annotation_col<-data.frame(c(colnames(top10_avgexp)))
colnames(annotation_col)<-"celltype"
rownames(annotation_col)<-colnames(top10_avgexp)
# heatmap_celltype_top10genes
# Fig S3E
p<-pheatmap(top10_avgexp,
            cluster_rows = F,
            cluster_cols = F,
            scale = 'row',
            cellwidth = 28,
            cellheight = 8,
            border_color="white",
            color = col3,
            # annotation_col = annotation_col,
            fontsize_col = 15,
            fontsize_row = 10
)

pdf("result/scRNA/TME/Figure5_Fibro_Macro/Figure_7_Fibro_cluster_heatmap.pdf",width = 10,height = 10)
print(p)
dev.off()
```
### scRNA Macro/Mono
```
Macro <- subset(PDAC_TME_sc, celltype == "Macro/Mono")
Macro <- RunPCA(Macro)
Macro <- RunHarmony(Macro,group.by.vars = "sample")
#PDAC_scRNA <- JackStraw(PDAC_scRNA, num.replicate = 10,reduction = "pca",assay = "RNA")
ElbowPlot(Macro,reduction = "harmony")

Macro <- RunUMAP(Macro,reduction = "harmony",dims = 1:30)
Macro <- FindNeighbors(Macro, reduction = "harmony", dims = 1:30)
Macro <- FindClusters(Macro,
                      resolution =0.2,
                      method ='igraph',
                      verbose = T)# 6
#defined celltype subtype
table(Macro$SCT_snn_res.0.2)
cluster<- c("Macro_c1",
            "Macro_c2",
            "Macro_c3",
            "Macro_c4",
            "Macro_c5",
            "Macro_c6",
            "Macro_c7",
            "Macro_c8"
)
new.cluster.ids <- cluster
names(new.cluster.ids)<- levels(Macro)
Macro <- RenameIdents(Macro, new.cluster.ids)
Macro@meta.data$celltype_subtype<-Macro@active.ident

table(Macro@active.ident)
Macro$site <- factor(Macro$site,levels = c("Normal pancreas",'Adjacent',"Liver metastases","PDAC primary"))
#UMAP plot
#Fig 3C
p<-DimPlot(Macro,group.by = "celltype_subtype",cols = col1[8:16],raster = T,split.by = "site")
pdf("/result/scRNA/TME/Figure5_Fibro_Macro/Figure_5_Macro_cellcluster_UMAP_legend.pdf",width = 21,height = 6)
print(p)
dev.off()

#Dotplot
#Fig S3H
markers <- c("CD68","CD163",'CD14','FCN1')
p <- DotPlot(Macro,features = markers,group.by = "celltype_subtype")+scale_color_gradient2(low = 'white',mid = col4[50],high = col4[20])+
  theme_bw()+theme(panel.grid.major = element_blank() ,axis.text.x = element_text(size = 15,color = "black",angle = 45,hjust = 1,vjust = 1),
                   axis.text.y = element_text(size = 15,color = "black"),
                   axis.title = element_text(size = 15,color = 'black'),
                   legend.text = element_text(size = 12,color = "black"))+labs(x="",y="")
p
pdf("/result/scRNA/TME/Figure5_Fibro_Macro/Figure_5_Macro_cellcluster_Marker.pdf",width = 6,height = 6)
print(p)
dev.off()
#percent
#Fig 3D
d7 <- dittoBarPlot(
  object =Macro ,
  var = "celltype_subtype",color.panel = col1[8:16],x.reorder = c(3,1,2,4),
  #colors = c(3,5,6,4,1,2),
  group.by = "site")
d7<-d7+theme(axis.text.x = element_text(size = 15,color = "black",angle = 45,hjust = 1,vjust = 1),
             axis.text.y = element_text(size = 15,color = "black"),
             axis.title = element_text(size = 15,color = 'black'),
             legend.text = element_text(size = 12,color = "black"))

pdf("result/scRNA/TME/Figure5_Fibro_Macro/Figure_6_Macro_percent_cellcluster.pdf",width = 4,height = 4)
print(d7)
dev.off()

marker_Macro <- FindAllMarkers(Macro,
                               assay = "SCT",
                               test.use = "wilcox",
                               logfc.threshold = 0.5,
                               #min.diff.pct = 0.3,
                               only.pos = T
)
write.table(marker_Macro,"result/scRNA/TME/Figure5_Fibro_Macro/marker_Macro.txt",sep = "\t",quote = F)

markers_cluster <- marker_Macro
top10<-markers_cluster %>% group_by(cluster) %>% top_n(n=10,wt=avg_log2FC)
head(top10)

gmt <- matrix("",12,length(cluster))
for (i in 1:length(cluster)){
  gmt[1,] <- cluster
  #gmt[2,] <-""
  gmt[3:12,i]<-top10$gene[(10*(i-1)+1):(10*i)]
  
}
gmt <- as.data.frame(t(gmt))
write.table(gmt,"analysis/Bulk-analysis/celltype_subtype/Macro_gmt.gmt",sep = "\t",row.names = F,col.names = F,quote = F)


top10<-top10[!duplicated(top10$gene),]
p<-DotPlot(object = Macro,
           features = top10$gene,
           assay = "SCT")+coord_flip()+theme(axis.text.x=element_text(angle=45,size=8))
pdata<-p$data
top10_avgexp<-pdata[,c(1,3,4)]
top10_avgexp<-na.omit(top10_avgexp)

top10_avgexp<-dcast(top10_avgexp,features.plot~id,value.var = 'avg.exp')
rownames(top10_avgexp)<-top10_avgexp$features.plot
top10_avgexp<-top10_avgexp[,-1]
top10_avgexp<-top10_avgexp[rowSums(top10_avgexp)>0,]
annotation_col<-data.frame(c(colnames(top10_avgexp)))
colnames(annotation_col)<-"celltype"
rownames(annotation_col)<-colnames(top10_avgexp)
# heatmap_celltype_top10genes
p<-pheatmap(top10_avgexp,
            cluster_rows = F,
            cluster_cols = F,
            scale = 'row',
            cellwidth = 28,
            cellheight = 8,
            border_color="white",
            color = col3,
            # annotation_col = annotation_col,
            fontsize_col = 15,
            fontsize_row = 10
)

pdf("result/scRNA/TME/Figure5_Fibro_Macro/Figure_7_Macro_cluster_heatmap.pdf",width = 10,height = 15)
print(p)
dev.off()

save(Fibro,Macro,file = "analysis/TME/Fibro_Macro.Mono_object.Rdata")
```
### Bulk RNA, calculate the correlation between Fibro subtypes and Macro/Mono subtypes.
### The method is the same as that used to calculate the main cell types, except that the gmt used to generate the top 10 marker genes is different.
```
#bulk RNA  cell type subtpe Fibro Mascro cor- pie plit#
bulkresult <-  read.table('analysis/Bulk-analysis/subtype_gmt/Fibro_all_cox.txt',sep = "\t",check.names = F,header = T)
bulkresult <-  bulkresult[,-1]
bulkresult <-  bulkresult[-grep("0",bulkresult$celltp1),]

result<-matrix(0,1*196*3,5)
colnames(result)<-c("Celltype1","Celltype2","Celltype1_celltype2","correlation","count")
result<-as.data.frame(result)

cell1<- unique(bulkresult$celltp1)
cell2<- unique(bulkresult$celltp2)


n=1
m=2
l=3
for (i in 1:length(cell1)) {
  tmp <- bulkresult[grep(cell1[i],bulkresult$celltp1),]
  for (j in 1:length(cell2)) {
    tmp2 <- tmp[grep(cell2[j],tmp$celltp2),]
    
    result[n,1] <- cell1[i]
    result[n,2] <- cell2[j]
    result[n,3] <- paste(cell1[i],"_",cell2[j],sep = "")
    result[n,4] <- "red"
    result[n,5] <-sum(tmp2 =="red")
    result[m,1] <- cell1[i]
    result[m,2] <- cell2[j]
    result[m,3] <- paste(cell1[i],"_",cell2[j],sep = "")
    result[m,4] <- "grey"
    result[m,5] <-sum(tmp2 =="grey")
    result[l,1] <- cell1[i]
    result[l,2] <- cell2[j]
    result[l,3] <- paste(cell1[i],"_",cell2[j],sep = "")
    result[l,4] <- "blue"
    result[l,5] <-sum(tmp2 =="blue")
    
    n=n+3
    m=n+1
    l=n+2
    
  }
  
}

result <- as.data.frame( result)
result$count <- as.numeric(result$count)
write.table(result,"analysis/bulk-analysis/subtype_gmt/Figure5_Fibro_Macro_correlation_gsva_bulkcorrelation_gsva_bulk.txt",sep = "\t",quote = F,row.names = F)

a <- unique(result$Celltype1) %>% rev()
#result$Celltype1 <-factor(result$Celltype1,levels = a)
result$Celltype2 <-factor(result$Celltype2,levels = a)

#Figure_celltype_correlation_gsva_bulk.pdf
#Fig3 E
p<-ggplot(data =result, aes(x = factor(1), y = count, fill = factor(correlation))) +
  geom_bar(stat="identity",position = "stack", color = NA) +
  scale_y_continuous(limits = c(0,12),expand=c(0,0)) +
  coord_polar("y") +
  facet_grid(Celltype2~Celltype1) +
  theme(axis.text=element_blank(),axis.title = element_blank(), 
        panel.background = element_blank(),panel.grid=element_blank(),
        legend.title=element_blank(), axis.ticks = element_blank(),
        strip.text.y = element_text(angle =0,hjust=0,color="black",size=8),strip.background = element_blank(),
        strip.text.x = element_text(color="black",size=7,angle = 90,vjust = 0),legend.text = element_text(size=10),
        legend.position = "bottom",panel.spacing  = unit(0.01, "lines")) +
  scale_fill_manual(limits=c("blue","grey","red"),values=c( "#377EB8","lightgray","#E41A1C"))

pdf("result/scRNA/TME/Figure5_Fibro_Macro/Figure5_Fibro_Macro_correlation_gsva_bulk.pdf",width=9,height = 9)
print(p)
dev.off()

#bulk Fibro c1 cor Macro c4 plot
filename <- list.files(path="analysis\\Bulk-analysis\\subtype_Fibro-Macro\\GSVA_correasubtype Fibro_Marco"
                       ,pattern="gsva.txt")

cox_res<-list()
for (i in 1:length(filename) ) {
  cox_res[[i]] <-read.table(paste("analysis\\Bulk-analysis\\subtype_Fibro-Macro\\GSVA_correasubtype Fibro_Marco\\",filename[i] ,sep = '')
                            ,sep = '\t')
}

##
filename2 <-filename
for (i in 1:length(filename)) {
  filename2[i] <- strsplit(filename[i],'_')[[1]][1]
  filename2[i] <- substr(filename2[i],6,20)
}

#Fig S4A
for (i in 1:length(cox_res)) {
  tmp <- cox_res[[i]]
  tmp <- as.data.frame(t(tmp))
  #Fibro cor Macro/mono
  colnames(tmp)
  p_tmp<- ggscatter( tmp ,x = 'Fibro_c1', y = 'Macro_c4',
                     add = "reg.line", conf.int = TRUE,color = col1[1],
                     add.params = list(fill = col1[20]),
                     ggtheme = theme_minimal()
  )+ stat_cor(method = "spearman",
              color='black',
              #p.accuracy = 0.001
  )+theme_bw()+theme(panel.grid.major = element_blank() ,axis.text.x = element_text(size = 12,color = "black",angle = 45,hjust = 1,vjust = 1),
                     axis.text.y = element_text(size = 12,color = "black"),
                     axis.title = element_text(size = 15,color = 'black'),
                     legend.text = element_text(size = 15,color = "black"))+labs(x="Fibroblast",y="Macro/Mono",title = filename2[i])
  pdf(paste("result/scRNA/TME/Figure5_Fibro_Macro/Fibro_c1_cor_macro_c4_all_bulk_plot/Figure_Cor_Fibro_c1_Macro_c4_",filename2[i],".pdf",sep = ""),width = 4,height = 4)
  print(p_tmp)
  dev.off()
}

```
### Calculate the 14 major hallmark enrichment by GSVA package
```
#Fibro
Idents(Fibro)
pbmc<-Fibro
expr <- AverageExpression(pbmc, assays = "RNA",slot = "data")[[1]]
expr <- expr[rowSums(expr)>0,]#ѡȡ????????
expr <- as.matrix(expr)
head(expr)

#14 major hallmark
genesets = readLines("cancer_pathway.gmt")
res <- strsplit(genesets, "\t")
names(res) <- vapply(res, function(y) y[1], character(1))
genesets <- lapply(res, "[", -c(1:2))
gsva.res <- gsva(expr, genesets, method="gsva") 
gsva.df <- data.frame(Genesets=rownames(gsva.res), gsva.res, check.names = F)
write.csv(gsva.df, "result/scRNA/TME/Figure5_Fibro_Macro/fibro_subtype_gsva_res_14major.csv", row.names = F)

#Fig S6B
p<-pheatmap(gsva.res,
            cluster_rows = T,
            cluster_cols = F,
            scale = 'row',
            cellwidth = 40,
            cellheight = 12,
            border_color="white",
            color = colorRampPalette(brewer.pal(10, "RdBu"))(50) %>% rev(),
            fontsize_col = 15,show_colnames = T,
            fontsize_row = 12
)
pdf(paste("result/scRNA/TME/Figure5_Fibro_Macro/Figure_fibro_subtype_gsva_res_14major.csv.pdf",sep = ""),width = 10,height = 10)
print(p)
dev.off()
#Macro
Idents(Macro)
pbmc<-Macro
expr <- AverageExpression(pbmc, assays = "RNA",slot = "data")[[1]]
expr <- expr[rowSums(expr)>0,]#ѡȡ????????
expr <- as.matrix(expr)
head(expr)

#14 major hallmark
genesets = readLines("D:\\空转\\ST_raw_data\\cancer_pathway.gmt")
res <- strsplit(genesets, "\t")
names(res) <- vapply(res, function(y) y[1], character(1))
genesets <- lapply(res, "[", -c(1:2))

gsva.res <- gsva(expr, genesets, method="gsva") 
gsva.df <- data.frame(Genesets=rownames(gsva.res), gsva.res, check.names = F)
write.csv(gsva.df, "result/scRNA/TME/Figure5_Fibro_Macro/macro_subtype_gsva_res_14major.csv", row.names = F)
#Fig S6C
p<-pheatmap(gsva.res,
            cluster_rows = T,
            cluster_cols = F,
            scale = 'row',
            cellwidth = 40,
            cellheight = 12,
            border_color="white",
            color = colorRampPalette(brewer.pal(10, "RdBu"))(50) %>% rev(),
            fontsize_col = 15,show_colnames = T,
            fontsize_row = 12
)
pdf(paste("result/scRNA/TME/Figure5_Fibro_Macro/Figure_Macro_subtype_gsva_res_14major.csv.pdf",sep = ""),width = 10,height = 10)
print(p)
dev.off()
```
# Spatial analysis of POSTN and SPP1 via BayesSpace
```

```
